pkg/cloudprovider/providers/azure/azure_loadbalancer.go

```diff
@@ -1006,35 +1006,46 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
 	}
 	expectedSecurityRules := []network.SecurityRule{}
 
-	if wantLb {
-		expectedSecurityRules = make([]network.SecurityRule, len(ports)*len(sourceAddressPrefixes))
-
+	if wantLb && len(sourceAddressPrefixes) > 0 {
+		expectedSecurityRules = make([]network.SecurityRule, len(ports))
 		for i, port := range ports {
 			_, securityProto, _, err := getProtocolsFromKubernetesProtocol(port.Protocol)
 			if err != nil {
 				return nil, err
 			}
-			for j := range sourceAddressPrefixes {
-				ix := i*len(sourceAddressPrefixes) + j
-				securityRuleName := az.getSecurityRuleName(service, port, sourceAddressPrefixes[j])
-				expectedSecurityRules[ix] = network.SecurityRule{
-					Name: to.StringPtr(securityRuleName),
-					SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
-						Protocol:                 *securityProto,
-						SourcePortRange:          to.StringPtr("*"),
-						DestinationPortRange:     to.StringPtr(strconv.Itoa(int(port.Port))),
-						SourceAddressPrefix:      to.StringPtr(sourceAddressPrefixes[j]),
-						DestinationAddressPrefix: to.StringPtr(destinationIPAddress),
-						Access:                   network.SecurityRuleAccessAllow,
-						Direction:                network.SecurityRuleDirectionInbound,
-					},
+			expectedSourceAddressPrefix, expectedSourceAddressPrefixes, err := az.getSourceAddressPrefix(sourceAddressPrefixes)
+			if err != nil {
+				return nil, err
+			}
+			var securityRuleName string
+			if useSharedSecurityRule(service) {
+				klog.V(2).Infof("Reconcile security group for service %s. When service is shared, the rule name includes sourceAddressPrefixes", service.Name)
```

Comment:
is it better put this `When service is shared, the rule name includes sourceAddressPrefixes` into code comment?